internal/stores/contracts.go

```diff
@@ -388,11 +371,24 @@ func (s *SQLStore) AddContract(c rhpv2.Contract) error {
 		NewUnlockHash:         c.Revision.NewUnlockHash,
 	}
 
-	// Insert contract.
-	return s.db.Create(&dbContractRHPv2{
-		ID:       fcid,
-		Revision: revision,
-	}).Error
+	return s.db.Transaction(func(tx *gorm.DB) error {
+		// Find host.
+		var host dbHost
+		err := s.db.Where(&dbHost{PublicKey: c.HostKey()}).
```

Comment:
So if I understand correctly a first or create is not possible here due to an FK constraint right? Kind of weird you can't insert two records with a constraint between them in one transaction.

Reply:
The public key of the host is not its primary key. So gorm has no way of knowing what host to add the contract to just from the provided public key. It would just create a new host and fail due to the unique constraint.